src/main/java/net/revelc/code/impsort/ImpSort.java

```diff
@@ -166,4 +178,55 @@ private static void convertAndAddImport(LinkedHashSet<Import> allImports, List<N
     allImports.add(imp);
   }
 
+  /*
+   * Extract all of the tokens from the main body of the file.
+   *
+   * This set of tokens represents all of the file's dependencies, and is used
+   * to figure out whether or not an import is unused.
+   */
+  private static Set<String> extractTokens(CompilationUnit unit) {
+    Set<String> tokens = new LinkedHashSet<String>();
+    NodeList<TypeDeclaration<?>> types = unit.getTypes();
+    types.forEach((Node node) -> {
+      Optional<TokenRange> tokenRange = node.getTokenRange();
+      if (tokenRange.isPresent() && tokenRange.get() != TokenRange.INVALID) {
+        JavaToken token = tokenRange.get().getBegin();
+        while (token != null && token != JavaToken.INVALID) {
+          String tokenString = token.toString();
+          if (!tokenString.isEmpty() && Character.isLetter(tokenString.charAt(0))) {
+            tokens.add(token.toString());
+          }
+          Optional<JavaToken> next = token.getNextToken();
```

Comment:
Is there a `hasNextToken` to loop on, in the while condition instead of pre-loading the token at the end of the previous loop?

Reply:
Unfortunately, no. But if I could convince you to move to a newer version of JavaParser, the TokenRange class gets an `Iterator<Token> iterator()` method, which would really simplify this code.

Reply:
Oh, I haven't been watching for updates. I think I used the latest that was available at the time. Currently, it is up to version 3.6.0. As long as it doesn't break anything and doesn't require a JDK newer than 8, feel free to bump it as needed in this pull request to simplify the parse code.